8327492: Remove applet usage and update DisposeInActionEventTest.html

alisenchung opened the pull request:
Root cause of the test failure was fixed with https://bugs.openjdk.org/browse/JDK-8316931, updating this test since the other fix also included a test update.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8327492](https://bugs.openjdk.org/browse/JDK-8327492): Remove applet usage and update DisposeInActionEventTest.html (**Bug** - P4)


### Reviewers
 * [Damon Nguyen](https://openjdk.org/census#dnguyen) (@DamonGuy - Committer) ⚠️ Review applies to [5f131ba1](https://git.openjdk.org/jdk/pull/17838/files/5f131ba1df8a8bd692f9dfc953b610014e44818b)
 * [Tejesh R](https://openjdk.org/census#tr) (@TejeshR13 - Committer) ⚠️ Review applies to [f185af12](https://git.openjdk.org/jdk/pull/17838/files/f185af128919903374393eca711da93bca2b214d)
 * [Abhishek Kumar](https://openjdk.org/census#abhiscxk) (@kumarabhi006 - Committer) ⚠️ Review applies to [f185af12](https://git.openjdk.org/jdk/pull/17838/files/f185af128919903374393eca711da93bca2b214d)
 * [Harshitha Onkar](https://openjdk.org/census#honkar) (@honkar-jdk - **Reviewer**) ⚠️ Review applies to [60bb2701](https://git.openjdk.org/jdk/pull/17838/files/60bb27016576ae4c002128e4af5f5c9bd2df9d26)
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**) ⚠️ Review applies to [0947ce09](https://git.openjdk.org/jdk/pull/17838/files/0947ce09d8ad755a6da49822c18ec4d4c18e9bfa)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17838/head:pull/17838` \
`$ git checkout pull/17838`

Update a local copy of the PR: \
`$ git checkout pull/17838` \
`$ git pull https://git.openjdk.org/jdk.git pull/17838/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17838`

View PR using the GUI difftool: \
`$ git pr show -t 17838`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17838.diff">https://git.openjdk.org/jdk/pull/17838.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17838#issuecomment-1942687029)

DamonGuy reviewed (commented):
I updated and ran the test changes. The instructions seem unclear. I double right click the icon in the icon tray and the icon temporarily disappears, then reappears. I think that's the correct behavior according to "After each action event the tray icon is removed from the tray and then added back in a second."

Also the instructions describe that there should be some "brief information about action events printed below". But I don't see any info printed.

Can you clear this up?

kumarabhi006 commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -21,76 +21,77 @@
+    public static void main(String[] args) throws Exception {
+        String instructions;
         traySupported = SystemTray.isSupported();
         if (!traySupported) {
```
`SystemTray.isSupported()` can be directly evaluated in if condition leading us to remove `traySupported` variable.

kumarabhi006 commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -21,76 +21,77 @@
-                    "Simply press PASS button."
-            };
+            instructions = "The test cannot be run because SystemTray is not supported.\n" +
+                           "Simply press PASS button.";
```
If systemTray is not supported, is it a good idea to just return and pass or skip the test ?

kumarabhi006 commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -21,76 +21,77 @@
-                    "  are printed, press PASS button else FAIL button."
-            };
+
+            instructions = "When the test starts, it adds the icon to the tray aread. If you\n" +
```
```suggestion
            instructions = "When the test starts, it adds the icon to the tray area. If you\n" +
```

alisenchung commented:
> I updated and ran the test changes. The instructions seem unclear. I double right click the icon in the icon tray and the icon temporarily disappears, then reappears. I think that's the correct behavior according to "After each action event the tray icon is removed from the tray and then added back in a second."
> 
> Also the instructions describe that there should be some "brief information about action events printed below". But I don't see any info printed.
> 
> Can you clear this up?

I've fixed the test outputs to go into the frame instead of sysout, so you should be able to see the information about the action events in the frame that pops up now. I've also updated the instructions to reflect this

DamonGuy commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -21,76 +21,78 @@
+ * @library ../../regtesthelpers /test/lib
+ * @build Sysout jtreg.SkippedException
+ * @run main/manual DisposeInActionEventTest
+ */
+public class DisposeInActionEventTest {
```
Add newline between these

DamonGuy commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -21,76 +21,78 @@
+    public static void main(String[] args) throws Exception {
+        String instructions;
+        if (!SystemTray.isSupported()) {
+            throw new jtreg.SkippedException("The test cannot be run because SystemTray is not supported.");
```
```suggestion
            throw new jtreg.SkippedException("The test cannot be run because " + 
            "SystemTray is not supported.");
```

DamonGuy commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -21,76 +21,78 @@
+                           "  don't see a tray icon, please, make sure that the tray area\n" +
+                           "  (also called Taskbar Status Area on MS Windows, Notification\n" +
+                           "  Area on Gnome or System Tray on KDE) is visible.\n" +
+                           "Double-click with " + clickInstruction + " button on the tray icon to trigger the\n" +
```
The instructions say to double click, but a single click right click is sufficient in macOS.

DamonGuy reviewed (approved):
Test LGTM now but I left some comments on minor issues.

TejeshR13 commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -21,76 +21,78 @@
+        String instructions;
+        if (!SystemTray.isSupported()) {
+            throw new jtreg.SkippedException("The test cannot be run because SystemTray is not supported.");
         } else {
```
Probably else is not required here, you can just remove it and put your code. Indentations can be reduced if you take out from else.

alisenchung commented:
/integrate

alisenchung commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -21,76 +21,73 @@
+        frame.setVisible(true);
 
-        System.setProperty("sun.awt.exception.handler", "DisposeInActionEventTest$EDTExceptionHandler");
+        System.setProperty("sun.awt.exception.handler\n",
```
i wasn't sure how to do the key value pairs for setProperty. Should there just be a space between sun.awt.exception.handler and DisposeInActionEventTest$EDTExceptionHandler?

honkar-jdk replied:
@alisenchung Why is System.setProperty("sun.awt.exception.handler\n", "DisposeInActionEventTest$EDTExceptionHandler") used ?

honkar-jdk commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -100,30 +97,30 @@ public void start() {
         trayIcon.setImageAutoSize(true);
         trayIcon.addActionListener(ev -> {
-            Sysout.println(ev.toString());
+            textArea.append(ev.toString());
```
Similar changes to other places where textArea.append() is used.

```suggestion
            textArea.append(ev.toString() + "\n");
```

honkar-jdk commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -21,76 +21,73 @@
-        if (!traySupported) return;
+        textArea = new JTextArea();
+        frame.getContentPane().add(textArea);
+        frame.setSize(200, 200);
```
Frame width can be increased to since the event details are not visible

honkar-jdk commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -21,76 +21,73 @@
+        String instructions = "When the test starts, it adds the icon to the tray area. If you\n" +
+                       "  don't see a tray icon, please, make sure that the tray area\n" +
+                       "  (also called Taskbar Status Area on MS Windows, Notification\n" +
+                       "  Area on Gnome or System Tray on KDE) is visible.\n" +
```
Instructions look more clear with the addition of newlines as below.

```suggestion
                       "  Area on Gnome or System Tray on KDE) is visible.\n\n" +
```

honkar-jdk commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -21,76 +21,73 @@
+                        clickInstruction + " the button on the tray icon to trigger the\n" +
+                       "  action event. Brief information about action events is printed\n" +
+                       "  in the frame. After each action event the tray icon is removed from\n" +
+                       "  the tray and then added back in a second.\n" +
```
```suggestion
                       "  the tray and then added back in a second.\n\n" +
```

honkar-jdk commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -21,76 +21,73 @@
+                .title("Test Instructions Frame")
+                .instructions(instructions)
+                .testTimeOut(10)
+                .rows(10)
```
Rows can be increased to 15.

```suggestion
                .rows(15)
```

alisenchung commented:
> @alisenchung Updated changes look good. I have one clarification though : Why is System.setProperty("sun.awt.exception.handler\n", "DisposeInActionEventTest$EDTExceptionHandler") used in this test?

The line was in the original test, but it looks like it's not needed so I'll remove it

aivanov-jdk commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -21,76 +21,69 @@
+                       "  don't see a tray icon, please, make sure that the tray area\n" +
+                       "  (also called Taskbar Status Area on MS Windows, Notification\n" +
+                       "  Area on Gnome or System Tray on KDE) is visible.\n\n" +
+                        clickInstruction + " the button on the tray icon to trigger the\n" +
```
_“Right-click the button”_ or _“Double-left click the button”_ doesn't make sense to me.

aivanov-jdk replied:
The recent update didn't clarify it, now the text reads _“Right-click the button”_ or _“Double click the button”_.

I assume _“the button”_ refers to the mouse button. But it's very confusing, it's not clear what is required; because “right-click” and “double-click” already define which _mouse button_ to use.

What you want to say is _“Right-click / Double-click on **the tray icon**…”_

aivanov-jdk commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -21,76 +21,69 @@
+                       "  action event. Brief information about action events is printed\n" +
+                       "  in the frame. After each action event the tray icon is removed from\n" +
+                       "  the tray and then added back in a second.\n\n" +
+                       "The test performs some automatic checks when removing the icon. If\n" +
+                       "  something is wrong the corresponding message is displayed below.\n" +
+                       "  Repeat clicks several times. If no 'Test FAILED' messages\n" +
+                       "  are printed, press PASS button else FAIL button.";
```
It's not displayed below, it's to the right, isn't it?

Fail the test automatically if you can detect the failure. You may show a message to the tester about the failure before shutting down the test.

aivanov-jdk replied:
This comment also implied amending the instructions to correspond to what the test does.

alisenchung replied:
the test doesn't actually do any automatic checks, so i've removed that line

aivanov-jdk commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -21,76 +21,69 @@
-        setVisible(true);
-        validate();
+        PassFailJFrame.builder()
+                .title("Test Instructions Frame")
```
More specific title would be better. Otherwise, you can remove `.title()` to get the same default title.

aivanov-jdk commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -21,76 +21,69 @@
 
-        System.setProperty("sun.awt.exception.handler", "DisposeInActionEventTest$EDTExceptionHandler");
+        textArea = new JTextArea();
+        frame.getContentPane().add(textArea);
```
```suggestion
        frame.getContentPane().add(new JScrollPane(textArea));
```
Put the text area into a scroll pane to allow viewing all text even it doesn't fit. And the messages with `ActionEvent` do not fit.

honkar-jdk replied:
@alisenchung The ActionEvent messages are not seen in the "Event Message Display" window with the current changes. 
The textArea should be added to JScrollPane as @aivanov-jdk suggested and not as `spane.add(textArea)`.

aivanov-jdk commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -21,76 +21,69 @@
+                .columns(45)
+                .testUI(DisposeInActionEventTest::showFrameAndIcon)
+                .build()
+                .awaitAndCheck();
```
I propose removing the icon from the system tray in a `finally` block. Otherwise, the test doesn't exit if you run it without jtreg.

alisenchung replied:
i'm not sure i understand this, the test is trying to remove right after the click action is delivered and not before the test is closing, so wouldn't putting the removal of the icon in a finally block defeat the purpose of the test?

honkar-jdk replied:
@alisenchung  Since the trayicon is added back after removal and PassFailJFrame handles only disposal of test windows, Aleksei is suggesting we handle the removal of trayicon explicitly something like below. You probably want to move the `systemTray` and `trayIcon` vars accordingly.

```
try {
    PassFailJFrame.builder()
                  .title("Event Message Display")
                  .instructions(instructions)
                  .testTimeOut(5)
                  .rows(15)
                  .columns(45)
                  .testUI(DisposeInActionEventTest::showFrameAndIcon)
                  .build()
                  .awaitAndCheck();
} finally {
     systemTray.remove(trayIcon);
}

```

aivanov-jdk replied:
That's correct! Otherwise, the orphan icon is left the system tray. When the test is run without jtreg, it doesn't terminate because the tray icon doesn't allow EDT to stop.

aivanov-jdk replied:
Alternatively, you can remove the system tray icon in `windowClosing` or `windowClosed` event of `WindowListener` associated with the frame.

aivanov-jdk commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -100,30 +93,30 @@ public void start() {
         trayIcon.setImageAutoSize(true);
         trayIcon.addActionListener(ev -> {
-            Sysout.println(ev.toString());
+            textArea.append(ev.toString() + "\n");
```
```suggestion
            textArea.append(ev + "\n");
```
`.toString()` is redundant.

aivanov-jdk commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -100,30 +93,30 @@ public void start() {
                     systemTray.add(trayIcon);
                 } catch (AWTException | InterruptedException e) {
-                    Sysout.println(e.toString());
-                    Sysout.println("!!! The test coudn't be performed !!!");
+                    textArea.append(e.toString() + "\n");
+                    textArea.append("!!! The test couldn't be performed !!!\\n");
                 }
```
Fail the test?

Print the stack trace of the exception with `e.printStackTrace()` — it would help analysing the failure.

aivanov-jdk commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -100,30 +93,30 @@ public void start() {
             systemTray.add(trayIcon);
         } catch (AWTException e) {
-            Sysout.println(e.toString());
-            Sysout.println("!!! The test coudn't be performed !!!");
+            textArea.append(e.toString() + "\n");
+            textArea.append("!!! The test couldn't be performed !!!\n");
         }
```
Also fail the test right away.

aivanov-jdk reviewed (changes requested):
The subject of the bug needs to be updated.

The root cause has been resolved under another bug. Here, you update the test… because it uses `Applet`. You don't actually fix the test to make it pass.

Perhaps even, create a _new bug_ with clear description and close [JDK-8325097](https://bugs.openjdk.org/browse/JDK-8325097) as duplicate of [JDK-8316931](https://bugs.openjdk.org/browse/JDK-8316931) because JDK-8316931 fixed both failures. For this reason, JDK-8316931 deserves being in the `@bug` tag in both tests: `DisposeInActionEventTest` and `ShowAfterDisposeTest`.

aivanov-jdk commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -62,12 +63,12 @@ public static void main(String[] args) throws Exception {
-                       "  something is wrong the corresponding message is displayed below.\n" +
+                       "  something is wrong the corresponding message is displayed.\n" +
                        "  Repeat clicks several times. If no 'Test FAILED' messages\n" +
                        "  are printed, press PASS button else FAIL button.";
```
My previous comment implies you should update the instructions to state that the test will _fail automatically_ if a failure condition detected.

After repeating clicking several times, the tester can press the _Pass_ button.

The tester should never click the _Fail_ button.

aivanov-jdk commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -81,8 +82,10 @@ private static JFrame showFrameAndIcon() {
         JFrame frame = new JFrame("DisposeInActionEventTest");
```
```suggestion
        JFrame frame = new JFrame("Event Message Display");
```
This is where events are displayed.

aivanov-jdk commented:
> The root cause has been resolved under another bug. Here, you update the test… because it uses `Applet`. You don't actually fix the test to make it pass.
> 
> Perhaps even, create a _new bug_ with clear description and close [JDK-8325097](https://bugs.openjdk.org/browse/JDK-8325097) as duplicate of [JDK-8316931](https://bugs.openjdk.org/browse/JDK-8316931) because JDK-8316931 fixed both failures. For this reason, JDK-8316931 deserves being in the `@bug` tag in both tests: `DisposeInActionEventTest` and `ShowAfterDisposeTest`.

Please, close this bug, [JDK-8325097](https://bugs.openjdk.org/browse/JDK-8325097), as duplicate of [JDK-8316931](https://bugs.openjdk.org/browse/JDK-8316931).

Create a new bug for updating the test with a clear subject. Then update the subject of this PR with the new bugid and summary.

Otherwise, as things stand now, you're creating _false_ records in JBS. The subject of this PR targeted for JDK-8325097 states *the `DisposeInActionEventTest.html` fails*. Then the PR resolves the failure. But the test **does _not_ fail** after JDK-8316931 is fixed.

honkar-jdk commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -21,76 +21,76 @@
+                    .testUI(DisposeInActionEventTest::showFrameAndIcon)
+                    .build()
+                    .awaitAndCheck();
+        } finally {
+            systemTray.remove(trayIcon);
         }
```
@alisenchung I missed adding null check here earlier. It is good to have this safety check to avoid NPE.

```suggestion
        } finally {
            if (systemTray != null) {
               systemTray.remove(trayIcon);
            }
         }
```

aivanov-jdk replied:
`SystemTray.getSystemTray()` can't return `null`. However, the static field `systemTray` can remain `null` in the test.

aivanov-jdk commented on test/jdk/java/awt/TrayIcon/ShowAfterDisposeTest/ShowAfterDisposeTest.java:
Does it make sense to remove the tray icon in this test as well when the test exits?

aivanov-jdk replied:
How is this resolved? You haven't updated `ShowAfterDisposeTest.java` to remove the tray icon when the test exists.

An answer of ‘No‘ is also valid. But you haven't said anything at all.

alisenchung replied:
oops, I misinterpreted what you wrote and updated the wrong file

alisenchung commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -21,76 +21,78 @@
+                       "  don't see a tray icon, please, make sure that the tray area\n" +
+                       "  (also called Taskbar Status Area on MS Windows, Notification\n" +
+                       "  Area on Gnome or System Tray on KDE) is visible.\n\n" +
+                        clickInstruction + " the button on the tray icon to trigger the\n" +
```
On windows you need to left click twice, on macos you need to right click once

aivanov-jdk replied:
[I did suggest clearer instructions](https://github.com/openjdk/jdk/pull/17838#discussion_r1506056726):
```java
                        clickInstruction + " on the tray icon to trigger the\n" +
```
You ignored it and marked resolved.

In the current instructions, _“the button”_ does not refer to anything obvious. What button? It's confusing. Remove _“the button”_ from the instructions, and the instructions are clear: _“Right-click / Double-click on **the tray icon**…”_

alisenchung replied:
I've also added left mouse button in parenthesis for windows, is it unnecessary? I think it should be clear now

alisenchung commented:
/integrate

aivanov-jdk commented on test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java:
```diff
@@ -65,8 +65,9 @@ public static void main(String[] args) throws Exception {
-                       "  Repeat clicks several times. Press PASS button.";
+                       "The test checks if any exceptions are thrown when removing and\n" +
+                       "  re-adding the icon. If something is wrong, the test will automatically fail.\n" +
+                       "  Repeat clicks several times Then press PASS button.";
```
You've lost the full stop after _“times”_. It's not worth a new PR.
